src/ydata_synthetic/synthesizers/timeseries/tscwgan/model.py

```diff
@@ -0,0 +1,248 @@
+
+    def c_lossfn(self, real_batch_, fake_batch_):
+        score_fake = self.critic(fake_batch_)
+        score_real = self.critic(real_batch_)
+        grad_penalty = self.gradient_penalty(real_batch_, fake_batch_)
+        c_loss = reduce_mean(score_fake) - reduce_mean(score_real) + grad_penalty
+        return c_loss
+
+    def g_lossfn(self, fake_batch_):
+        score_fake = self.critic(fake_batch_)
+        g_loss = - reduce_mean(score_fake)
+        return g_loss
+
+    def _make_fake_batch(self, real_batch, noise_batch):
+        """Generate a batch of fake records and return it with the batch of used conditions.
+        Conditions are the first elements of records in the real batch."""
+        cond_batch = real_batch[:, :self.cond_dim]
+        gen_input = concat([cond_batch, noise_batch], axis=1)
+        return self.generator(gen_input, training=True), cond_batch
+
+    def gradient_penalty(self, real, fake):
+        gp = gradient_penalty(self.critic, real, fake, mode=Mode.DRAGAN)
+        return gp
+
+    def _generate_noise(self):
+        "Gaussian noise for the generator input."
+        while True:
+            yield normal(size=self.noise_dim)
+
+    def get_batch_noise(self):
+        "Create a batch iterator for the generator gaussian noise input."
+        return iter(tfdata.Dataset.from_generator(self._generate_noise, output_types=float32)
+                                .batch(self.batch_size)
+                                .repeat())
+
+    def get_batch_data(self, data, n_windows= None):
+        if not n_windows:
+            n_windows = len(data)
+        data = reshape(convert_to_tensor(data, dtype=float32), shape=(-1, self.data_dim))
+        return iter(tfdata.Dataset.from_tensor_slices(data)
+                                .shuffle(buffer_size=n_windows)
+                                .batch(self.batch_size).repeat())
+
+    def sample(self, cond_array, n_samples):
+        """Provided that cond_array is passed, produce n_samples for each condition vector in cond_array."""
+        assert len(cond_array.shape) == 2, "Condition array should have 2 dimensions."
+        assert cond_array.shape[1] == self.cond_dim, \
+            f"Each sequence in the condition array should have a {self.cond_dim} length."
+        n_conds = cond_array.shape[0]
+        steps = n_samples // self.batch_size + 1
+        data = []
+        z_dist = self.get_batch_noise()
+        for seq in range(n_conds):
+            cond_seq = expand_dims(convert_to_tensor(cond_array.iloc[seq], float32), axis=0)
+            cond_seq = tile(cond_seq, multiples=[self.batch_size, 1])
+            for step in trange(steps, desc=f'Synthetic data generation - Condition {seq+1}/{n_conds}'):
+                gen_input = concat([cond_seq, next(z_dist)], axis=1)
+                records = make_ndarray(make_tensor_proto(self.generator(gen_input, training=False)))
+                data.append(records)
+        return DataFrame(concat(data, axis=0))
```

Comment:
@jfsantos-ds
Why are we concatenating the samples? 
And why are we assuming several conditions?

Reply:
Changed, if accepting multiple condition sets the samples from different conditions shouldn't be concatenated.
Reverting to one array for a single condition set input.